subprojects/core/src/main/java/org/gradle/internal/operations/notify/BuildOperationNotificationBridge.java

```diff
@@ -26,27 +29,58 @@
 import org.slf4j.LoggerFactory;
 
 import javax.annotation.Nullable;
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 
-class BuildOperationNotificationBridge implements BuildOperationNotificationListenerRegistrar, Stoppable {
+public class BuildOperationNotificationBridge implements BuildOperationNotificationListenerRegistrar, Stoppable {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(BuildOperationNotificationBridge.class);
 
     private Listener operationListener;
     private final BuildOperationListenerManager buildOperationListenerManager;
+    private BuildOperationRecorder buildOperationRecorder;
 
-    BuildOperationNotificationBridge(BuildOperationListenerManager buildOperationListenerManager) {
+    BuildOperationNotificationBridge(BuildOperationListenerManager buildOperationListenerManager, BuildOperationRecorder buildOperationRecorder) {
         this.buildOperationListenerManager = buildOperationListenerManager;
+        this.buildOperationRecorder = buildOperationRecorder;
+    }
+
+    public void start(GradleInternal gradle) {
+        buildOperationRecorder.start();
+
+        // ensure we only store events for configuration phase to keep overhead small
+        // when build scan plugin is not applied
+        gradle.rootProject(new Action<Project>() {
+            @Override
+            public void execute(Project project) {
+                project.afterEvaluate(new Action<Project>() {
+                    @Override
+                    public void execute(Project project) {
+                        BuildOperationNotificationBridge.this.buildOperationRecorder.discardEventsAndStop();
+                    }
+                });
+            }
+        });
     }
 
     @Override
     public void registerBuildScopeListener(BuildOperationNotificationListener notificationListener) {
         if (operationListener == null) {
             operationListener = new Listener(notificationListener);
+            List<BuildOperationRecorder.RecordedBuildOperation> recordedBuildOperations = buildOperationRecorder.retrieveEventsAndStop();
+            for (BuildOperationRecorder.RecordedBuildOperation storedEvent : recordedBuildOperations) {
+                BuildOperationDescriptor buildOperationDescriptor = storedEvent.buildOperation;
+                Object event = storedEvent.event;
+                if (storedEvent.eventType == BuildOperationRecorder.RecordedBuildOperation.OperationEventType.START) {
+                    operationListener.started(buildOperationDescriptor, (OperationStartEvent) event);
+                }else if (storedEvent.eventType == BuildOperationRecorder.RecordedBuildOperation.OperationEventType.FINISHED) {
+                    operationListener.finished(buildOperationDescriptor, (OperationFinishEvent) event);
+                }
+            }
             buildOperationListenerManager.addListener(operationListener);
```

Comment:
This mechanism isn't safe, and could lose notifications that come in while we are replaying the recorded ones. One option would be to add the new listener before removing the recording one (and block/record any notifications that appear while we are replaying).

Probably a better way is to consolidate the 2 `BuildOperationListener` implementations into 1. This listener would record notifications until a `BuildOperationNotificationListener` is added, and the directly forward notifications once the listener is added. We'd need to coordinate so that we continue to record while we are replaying, with a thread-safe switch to direct forwarding of notifications.